vllm/core/scheduler.py

```diff
@@ -1058,13 +1061,16 @@ def free_seq(self, seq: Sequence) -> None:
         self.block_manager.free(seq)
 
     def free_finished_seq_groups(self) -> None:
-        for queue in [self.running, self.swapped, self.waiting]:
-            self._finished_requests_ids += [
-                seq_group.request_id for seq_group in queue
-                if seq_group.is_finished()
-            ]
-        self.running = deque(seq_group for seq_group in self.running
-                             if not seq_group.is_finished())
+        # finished requests in self.waiting and self.swapped are already
+        # appended to self._finished_requests_ids during the scheduling.
+        # the only new finished requests are in self.running.
+        remaining: Deque[SequenceGroup] = deque()
+        for seq_group in self.running:
+            if seq_group.is_finished():
+                self._finished_requests_ids.append(seq_group.request_id)
+            else:
+                remaining.append(seq_group)
+        self.running = remaining
```

Comment:
@youkaichao how about wrapping this in
```python
 if any(seq_group.is_finished() for seq_group in self.running):
 ```
 to avoid rebuilding the deque every time, since it requests will be finished relatively rarely.

Reply:
I plan to refactor this into dictionary in the future, so that we can easily delete requests.